### PR TITLE
feat: add lightweight CNN divergence diagnostics (#835)

### LINKED ISSUE #835: bug: investigate LightweightCNNExtractor divergence at ≥28K training steps
## Context

Issue #193 (PR #833) showed `LightweightCNNExtractor` (preset: `lightweight_cnn`) diverging badly
at the final eval checkpoint:

- best eval reward at 12 K steps: −4.20
- final eval reward at 32 K steps: **−64.09** (catastrophic drop)

Config: `configs/scenarios/multi_extractor_eval_193.yaml`, seed 42, RTX 3070, 32 K PPO steps.

Full data: `docs/context/issue_193_feature_extractor_evaluation.md`

## Hypothesis

The extractor produces 272 output features — the largest feature dim of all 4 extractors tested.
This high-dimensional output may cause gradient norm spikes in the shared policy head when combined
with the PPO default learning rate. Possible causes:

1. Learning-rate vs feature-scale mismatch.
2. Missing batch norm / feature normalisation after the CNN output.
3. PPO clip ratio interacting poorly with the larger gradient surface.

## Scope

- Reproduce the divergence deterministically (rerun same config, verify same curve shape).
- Instrument gradient norms and feature statistics during the divergence window (20–32 K steps).
- Try learning-rate halving and/or feature normalisation as candidate fixes.
- Report whether the extractor can be stabilised or should be marked as experimental.

## Definition of Done

- [ ] Divergence reproduced and confirmed as systematic (not a one-seed fluke).
- [ ] Root cause identified (gradient norm / scale / architecture).
- [ ] Either a fix is verified stable, or the extractor is documented as experimental with a
  clear failure condition.

## Related

- Evaluated in: #193, PR #833
- Extractor code: `robot_sf/feature_extractors/lightweight_cnn_extractor.py`

## Goal / Problem

**Task type**
- bug investigation / experimental extractor triage

**Context**
- `lightweight_cnn` diverged hard in the issue-193 study, but it is not the current promotion candidate.
- The goal here is to decide whether the divergence is a real, fixable trainability defect or whether the extractor should simply remain documented as experimental and de-prioritized.

## Added Value Estimation

- User value: Low-medium. This only matters if the CNN extractor remains a plausible future option.
- Benchmark value: Low-medium. It does not currently block the preferred feature-extractor path.
- Maintenance value: Medium. A clear verdict avoids recurring investigation of the same unstable branch.
- Why now: only useful as a bounded triage pass so the team can stop treating the divergence as an open mystery.

## Effort Estimation

- Rough estimate (hours): 6-10
- Best estimate (hours): 8
- Unknowns:
  - whether the divergence reproduces consistently across reruns
  - whether a fix is config-level or requires architecture changes

## Complexity Estimation

- Implementation complexity: Moderate.
- Dependencies:
  - issue `#193` artifacts
  - PPO training instrumentation
  - the CNN extractor implementation and candidate normalisation hooks
- Open questions:
  - is the failure mostly a learning-rate/scale mismatch, or is the architecture simply a poor fit here?

## Risk Assessment

- Functional risk: Low for the repo overall because this path is not the current default.
- Compatibility risk: Low.
- Rollout risk: Medium because confirming the root cause may require reruns and instrumentation.
- Mitigation: keep the scope to deterministic reproduction plus one narrow mitigation hypothesis before any broader sweep.

## Affected Files

- `robot_sf/feature_extractors/lightweight_cnn_extractor.py`
- PPO training/eval config surfaces used in issue `#193`
- `docs/context/issue_193_feature_extractor_evaluation.md`
- any focused training instrumentation added during the triage

## Success Metrics

- The issue ends with a clear verdict: stabilizable with a bounded fix, or experimental and not worth pursuing now.
- The divergence is no longer treated as a vague open question.
- Any proposed mitigation stays narrow and reproducible.

## Validation / Testing

- [ ] Reproduce the divergence with the original issue-193 config path.
- [ ] Capture the gradient/feature diagnostics needed for the chosen mitigation hypothesis.
- [ ] Re-run one bounded mitigation check or document why the path should be demoted instead.

## Estimate Discussion

- Why these values were proposed:
  - This is a real investigation, but it is no longer on the critical path for default or promotion decisions.
- Uncertainty / confidence:
  - Medium confidence in the observed failure mode, lower confidence in the best fix.
- What evidence would move the estimate:
  - If the issue reproduces with one obvious config-level cause, value and success probability increase.
  - If the architecture still fails after one narrow mitigation, priority should drop further.

## Project Metadata

- Priority: Low
- Effort (h): 8
- Reviewed: 2026-04-28

### PR DESCRIPTION
**Pull request by ll7**
## Summary
Adds an opt-in PPO diagnostics path for `lightweight_cnn` triage and records the issue #835 32 K rerun verdict: the issue-193 catastrophic final drop did not reproduce on the RTX 3070 rerun.

## Linked Issues
- Closes #835
- Relates to #193

## What Changed
- Added `DiagnosticPPO` to capture PPO gradient norms at clipping time and optional feature statistics from `LightweightCNNExtractor`.
- Added `configs/scenarios/lightweight_cnn_issue_835_repro.yaml` as the focused 32 K reproduction config with diagnostics enabled from 20 K onward.
- Kept diagnostics opt-in in `scripts/multi_extractor_training.py`, so normal multi-extractor runs still use plain SB3 `PPO`.
- Added focused tests for feature-stat telemetry, diagnostic helper behavior, JSONL writing, and the PPO smoke path.
- Added `docs/context/issue_835_lightweight_cnn_divergence.md` and linked it from `docs/context/README.md`.

## Why It Matters
- Added value: turns the `lightweight_cnn` divergence from a vague open question into a reproducible diagnostic workflow plus a documented bounded verdict.
- Expected impact: reviewers can rerun the exact config and inspect `training_diagnostics.jsonl` without changing default training behavior.
- Why this is worth merging now: it prevents repeated investigation of the issue-193 cliff while preserving the extractor as experimental rather than promoting or removing it prematurely.

## Validation / Proof
- Commands run:
  - `uv run pytest tests/test_ppo_diagnostics.py tests/test_feature_extractors.py::TestFeatureExtractors::test_lightweight_cnn_extractor_forward -q`
  - `ROBOT_SF_MULTI_EXTRACTOR_TEST_MODE=1 uv run python scripts/multi_extractor_training.py --config configs/scenarios/lightweight_cnn_issue_835_repro.yaml --output-root tmp/issue_835_test_mode`
  - `uv run python scripts/multi_extractor_training.py --config configs/scenarios/lightweight_cnn_issue_835_repro.yaml --output-root tmp/issue_835_repro --verbose`
  - `BASE_REF=origin/main scripts/dev/pr_ready_check.sh`
- Evidence that the change works here:
  - Full readiness gate: `2967 passed, 18 skipped, 3 warnings`; changed-file coverage was 100% for `robot_sf/feature_extractors/lightweight_cnn_extractor.py` and `robot_sf/training/ppo_diagnostics.py`.
  - 32 K RTX 3070 rerun completed successfully in 401.1 s.
  - Eval curve did not reproduce the `-64.09` cliff: best mean reward `3.79` at 20 K, final mean reward `-8.37` at 32 K.
  - Diagnostics from 20 K onward showed stable feature scale and no late gradient spike matching the original collapse.

## Risks / Rollout
- Compatibility risks: low; diagnostics are opt-in via `run.training_diagnostics`, and `LightweightCNNExtractor` feature stats are opt-in via `record_feature_stats`.
- Failure modes: diagnostics add host synchronization when feature stats are enabled, so they should stay off for production-scale runs unless needed for triage.
- Rollback or fallback plan: use the existing non-diagnostic multi-extractor configs, or disable `training_diagnostics` and `record_feature_stats` in the issue config.

## Docs / Provenance
- Updated docs: `docs/context/issue_835_lightweight_cnn_divergence.md`, `docs/context/README.md`.
- Relevant design or provenance notes: issue #193 remains the source of the original observed cliff; issue #835 records the bounded non-reproduction rerun.
- Assumption preserved: a single 32 K rerun is enough to reject deterministic reproduction, not enough to promote `lightweight_cnn`.

## Follow-Up Issues
- Deferred work: longer multi-seed feature-extractor comparison remains the useful next evidence, but it belongs to the broader feature-extractor campaign rather than this issue.
- Issues opened for follow-up: none; this PR documents the boundary instead of creating new scope.

## Reviewer Notes
- Verify that diagnostics remain opt-in for regular multi-extractor runs.
- The extractor should remain experimental; this PR does not claim it is stable or promotion-ready.

### CODE REVIEW

**Comment by ll7** [comment]
Addressed the current inline review feedback in `37aaef61`.

Changes:
- batched `LightweightCNNExtractor` feature-stat collection into one device-to-host sync
- accumulated gradient L2 norms on-device before a single sync in `ppo_diagnostics`
- wrapped the `clip_grad_norm_` patch in a scoped context manager instead of mutating it inline
- added a focused test covering restoration of the original `clip_grad_norm_`

Validation:
- `uv run pytest tests/test_ppo_diagnostics.py tests/test_feature_extractors.py::TestFeatureExtractors::test_lightweight_cnn_extractor_forward -q`
- `BASE_REF=origin/main scripts/dev/pr_ready_check.sh`
- Full gate result: `2969 passed, 17 skipped`
